crates/typst-syntax/src/ast.rs

```diff
@@ -220,64 +250,68 @@ impl<'a> Expr<'a> {
-            SyntaxKind::Equation => node.cast().map(Self::Equation),
-            SyntaxKind::Math => node.cast().map(Self::Math),
-            SyntaxKind::MathText => node.cast().map(Self::MathText),
-            SyntaxKind::MathIdent => node.cast().map(Self::MathIdent),
-            SyntaxKind::MathShorthand => node.cast().map(Self::MathShorthand),
-            SyntaxKind::MathAlignPoint => node.cast().map(Self::MathAlignPoint),
-            SyntaxKind::MathDelimited => node.cast().map(Self::MathDelimited),
-            SyntaxKind::MathAttach => node.cast().map(Self::MathAttach),
-            SyntaxKind::MathPrimes => node.cast().map(Self::MathPrimes),
-            SyntaxKind::MathFrac => node.cast().map(Self::MathFrac),
-            SyntaxKind::MathRoot => node.cast().map(Self::MathRoot),
-            SyntaxKind::Ident => node.cast().map(Self::Ident),
-            SyntaxKind::None => node.cast().map(Self::None),
-            SyntaxKind::Auto => node.cast().map(Self::Auto),
-            SyntaxKind::Bool => node.cast().map(Self::Bool),
-            SyntaxKind::Int => node.cast().map(Self::Int),
-            SyntaxKind::Float => node.cast().map(Self::Float),
-            SyntaxKind::Numeric => node.cast().map(Self::Numeric),
-            SyntaxKind::Str => node.cast().map(Self::Str),
-            SyntaxKind::CodeBlock => node.cast().map(Self::Code),
-            SyntaxKind::ContentBlock => node.cast().map(Self::Content),
-            SyntaxKind::Parenthesized => node.cast().map(Self::Parenthesized),
-            SyntaxKind::Array => node.cast().map(Self::Array),
-            SyntaxKind::Dict => node.cast().map(Self::Dict),
-            SyntaxKind::Unary => node.cast().map(Self::Unary),
-            SyntaxKind::Binary => node.cast().map(Self::Binary),
-            SyntaxKind::FieldAccess => node.cast().map(Self::FieldAccess),
-            SyntaxKind::FuncCall => node.cast().map(Self::FuncCall),
-            SyntaxKind::Closure => node.cast().map(Self::Closure),
-            SyntaxKind::LetBinding => node.cast().map(Self::Let),
-            SyntaxKind::DestructAssignment => node.cast().map(Self::DestructAssign),
-            SyntaxKind::SetRule => node.cast().map(Self::Set),
-            SyntaxKind::ShowRule => node.cast().map(Self::Show),
-            SyntaxKind::Contextual => node.cast().map(Self::Contextual),
-            SyntaxKind::Conditional => node.cast().map(Self::Conditional),
-            SyntaxKind::WhileLoop => node.cast().map(Self::While),
-            SyntaxKind::ForLoop => node.cast().map(Self::For),
-            SyntaxKind::ModuleImport => node.cast().map(Self::Import),
-            SyntaxKind::ModuleInclude => node.cast().map(Self::Include),
-            SyntaxKind::LoopBreak => node.cast().map(Self::Break),
-            SyntaxKind::LoopContinue => node.cast().map(Self::Continue),
-            SyntaxKind::FuncReturn => node.cast().map(Self::Return),
+            SyntaxKind::Linebreak => Some(Self::Linebreak(Linebreak(node))),
```

Comment:
In a way, this is sort of a downgrade in safety since we're skipping the `from_untyped` implementation now, so we could accidentally pass the wrong node to a particular Expr variant, but in another, I guess there are less redundant steps. An alternative to guarantee there won't be a mismatch but also improve performance would be to have a separate macro that generates each line in the match, but that could be discussed.

An alternative is to instead create an ad-hoc macro to generate unit tests which guarantee the casting works correctly.

Reply:
I also had that thought, but it's true that it's also a performance improvement (unless the compiler managed to optimize that away before). As it's not too hard to check that it's that correct, I think it's ok.